[lsyurea] iP

lsyurea opened the pull request:
# DukePro

> "Your mind is for having ideas, not holding them." – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- **Text-based**
- **Easy to learn**
- **FAST SUPER FAST** to use

All you need to do is,

1. [Download it from here](https://github.com/lsyurea/ip/releases/download/A-Jar/duke.jar).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉
   
And it is **FREE**!

## Features:

- [x] Managing tasks
- [x] Managing deadlines
- [ ] ~~Reminders (coming soon)~~

If you're a Java programmer, you can use it to practice Java too. Here's the main method in Duke:

```java
public class Main {
    public static void main(String[] args) {
        Ui.greet();
        runProgram();
        Ui.exit();
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

lsyurea committed:
Update gitignore file

- add *.class to prevent class file from being committed after compilation

lsyurea committed:
Update Duke.java

-Add greet functionality
-Add exit functionality
-Refactor chatbot name

lsyurea committed:
Update Duke.java

- Add echo functionality to echo user input

lsyurea committed:
Update Duke.java

- Add adding functionality of task
- Add list functionality to list all added tasks

lsyurea committed:
Add Mark as Done

- Remove echo functionality as it is not needed anymore
- Added Task class to manage the task being marked or unmarked
- Added TaskList class to manage tasks

lsyurea committed:
Add ToDos, Events and Deadlines

-Change Task to abstract
-Add ToDoTask which extends Task
-Add EventTask which extends Task and takes in from and to dates
-Add Deadlines which extends Task and take in by date

lsyurea committed:
Update TaskList.java

-Changed string representation of adding tasks

lsyurea committed:
Added Automated Text UI Testing

-Change runtest.sh to be executable
-Update input.txt to include more test cases
-Update EXPECTED.TXT to reflect the  UI changes

lsyurea committed:
Added Error Handling

-Added DukeException
-Added empty description error handling and wrong command error handling
-Update text-ui-test to test for error handling

lsyurea committed:
Added Delete Functionality

-Changed from array to arraylist collections
-Updated EXPECTED.TXT to reflect input.txt change
-Updated input.txt to test for delete
-Updated Duke.java to handle delete

lsyurea committed:
Added Enums

-Add enum for TaskTypes
-Add enum for ModifyTask

lsyurea committed:
Update Duke.java

Change spacing of switch statements

switch statements need to follow the correct coding standard and indentation.

lsyurea committed:
Update .gitignore

Add gitignore for vscode

Prevents the vscode configuration from being updated to github.

lsyurea committed:
Update .gitignore

Added data/

Prevents sensitive data such as your tasks from being updated on github.

lsyurea committed:
Add Storage.java

Add saveTask

Saves serialisable objects to the specified txt file in the filepath.

Add loadTask

Loads serialisable objects from specified txt file in the filepath.

lsyurea committed:
Update Task.java

Add Task to implement Serializable

To allow Task object to be saved as objects.

lsyurea committed:
Update TaskList.java

Add saveTask method from Storage

To allow task to be loaded to the txt file every time the task changes.

Add loadTask method from Storage

To loadTask the first time taskList is instantiated.

lsyurea committed:
Merge branch 'branch-Level-7'

Add Storage feature to main branch

This allow a txt file to store all inputs for data persistence.

lsyurea committed:
Update DeadlineTask.java

Change date to a localdate object

Allow the customisation of what datestring to print out.

lsyurea committed:
Update DeadlineTask.java

Change string representation

Previously returns a formatter instead of the date string that uses the formatter.

lsyurea committed:
Merge branch 'branch-level-8'

Add parsed date to the deadline task

This allows the date format printed to be correct.

lsyurea committed:
Update Duke.java

Remove code and shift code

The codes that are more relevant to command parsing are placed in the Parser.java. The codes that are more relevant in Ui are placed in the Ui.java.

lsyurea committed:
Create Parser.java

Added addTask, modifyTask and parseCommands

Used to take in a string input that is not bye and executes the correct command.

lsyurea committed:
Update Storage.java

Implement Ui

Use Ui functions instead

lsyurea committed:
Update TaskList.java

Use Ui functions instead

Instead of handing the print statements, Ui functions are used.

lsyurea committed:
Create Ui.java

Create a list of Ui

Prints out the relevant instructions for each function.

lsyurea committed:
Add packages

Move all files into duke package

Move files into task package

All Deadline, Event, Task, TaskList and Todo files are stored in task package. This is to organise files.

lsyurea committed:
Merge branch 'add-gradle-support'

lsyurea committed:
Add junit Test

Test EventTaskTest

Added test for toString and markTaskAsDone methods.

Test TodoTaskTest

Added test for toString and markTaskAsDone methods.

lsyurea committed:
Add Find

Allow User to search for the task keyword

Updated UI.java to include the correct print statements. Updated TaskList to implement the search logic. Updated Parser.java to include the find command.

lsyurea committed:
Add javadocs

Add javadocs for all files in duke package

This ensure that there is proper documentation.

lsyurea committed:
Update TaskList.java

Ensure coding standards are adhered to

duke.txt changed to duke.ser for proper file extension in the storing of serializable objects. Secondary checks on all files such that indentation is correct and imports alignment is correct.

lsyurea committed:
Merge pull request #1 from lsyurea/branch-A-CodingStandard

Update TaskList.java

Ensures that ser files are used instead of txt files which is more suited to store serialized objects. Ensure that coding standards such as spacing and import order are adhered to as well.

lsyurea committed:
Merge pull request #2 from lsyurea/branch-A-JavaDoc

Add javadocs

lsyurea committed:
Merge pull request #3 from lsyurea/branch-Level-9

Add Find

lsyurea committed:
Update Storage.java

Give reason for casting

lsyurea committed:
Fix Deadline Error

Update to handle error caused by the wrong string format for date.

lsyurea committed:
Update TaskList.java

Update the javadocs of the function to findTasks in Level 9.

lsyurea committed:
Update build.gradle

Adds the relevant dependencies for javaFX

lsyurea committed:
Add checkstyle

Update build.gradle

This is to update the dependencies for checkstyle.

Add checkstyle.xml

This is to add checkstyle for java.

Add suppressions.xml

This is to check for suppresswarnings in java.

lsyurea committed:
Update checkstyle

Modify files

This is to adhere to the javadocs standard for checkstyle.

lsyurea committed:
Update build.gradle

Added the relevant dependencies

lsyurea committed:
Add fxml files

Needed for javaFX DialogBox and MainWindow.

lsyurea committed:
Add images

Images for the chatbot

lsyurea committed:
Add javaFX support

Update the files to integrate javaFX into it.

lsyurea committed:
Merge pull request #4 from lsyurea/branch-Level-10

Branch level 10

lsyurea committed:
Resolve Conflict

Resolve merge conflicts.

lsyurea committed:
Update TaskList.java

Merge conflicts.

lsyurea committed:
Fix checkstyle issues

Fixed all issues pertaining to spaces and checkstyles.

lsyurea committed:
Update dukeImage

Change dukeImage to be the same as userImage because of unsupported file type.

lsyurea committed:
Update Parser.java

Added exit to close the GUI.

lsyurea committed:
Update Parser.java

Returns the string to pass checkstyle.

lsyurea committed:
Update javadocs

Updates the outdated UI javadocs.

tjch-o commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,155 @@
+        } else {
+            String[] inputSplit = input.split(" ");
+            try {
+                switch (inputSplit[0]) {
```
good use of a switch here!

tjch-o commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,74 @@
+            FileInputStream fis = new FileInputStream(this.filePath);
+            ObjectInputStream ois = new ObjectInputStream(fis);
+            // Object is of type ArrayList<Task> hence it is possible to case directly
+            @SuppressWarnings("unchecked")
```
wow good use of suppresswarnings and also using `FileInputStream`! would have never thought about this approach...

tjch-o commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,155 @@
+     * @param input The user input.
+     * @throws DukeException If the user input is invalid.
+     */
+    private static String addTask(TaskType tsk, String input) throws DukeException {
```
should `tsk` be renamed to "task" for better readability?

lsyurea replied:
I agree that it should be named task. Thanks :)

tjch-o commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,155 @@
+                return Ui.returnErrorString(new DukeException("The description of a deadline cannot be empty."));
+            } catch (ArrayIndexOutOfBoundsException e) {
+                return Ui.returnErrorString(new DukeException("The task of a deadline cannot be empty."));
+            } catch (java.time.format.DateTimeParseException e) {
```
good that you handled the invalid datetime format case!

tjch-o reviewed (commented):
LGTM! the code was really well written and I did not find much issues with it.

SungMatt commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,155 @@
+     * @param tsk The type of task to be added.
+     * @param input The user input.
+     * @throws DukeException If the user input is invalid.
+     */
```
What does this return?

SungMatt commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,155 @@
+     * Modifies a Task object in the specified list of tasks.
+     * @param mod The type of modification to be made.
+     * @param input The user input.
+     * @throws DukeException If the user input is invalid.
```
What does this return too!

SungMatt commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,74 @@
+     */
+    public ArrayList<Task> loadTask() {
+        try {
+            FileInputStream fis = new FileInputStream(this.filePath);
```
Nice streams!

SungMatt commented on src/main/java/duke/task/EventTask.java:
```diff
@@ -0,0 +1,32 @@
+     */
+    public EventTask(String description, String fromDate, String toDate) {
+        super(description);
+        this.fromDate = fromDate;
```
Is this supposed to be of a LocalDate too?

SungMatt commented on text-ui-test/input.txt:
```diff
@@ -0,0 +1,12 @@
+todo borrow books
```
Great comprehensive test!!

SungMatt reviewed (approved):
LGTM! Just a few javadocs that could use a little more touchup, and you should be good!

lsyurea committed:
Update javadocs

In this commit, the Javadocs in the four files have been updated to improve code documentation and clarity. The changes include:
- 'DukeException.java': Changed string representation to use only ASCII characters.
-'Parser.java': Updated Javadocs to include missing return parameter.
-'EventTask.java': Added missing return parameter.
-'TaskList.java': Added missing return parameter.

These updates enhance code understanding and maintainability.

lsyurea committed:
Add assertions

In this assertions I have added input validation:
- "Storage.java": filepath should not be empty and tasks should not be null when saving something.
-"TaskList.java": db must not be null when TaskList() is created.

This ensures that the functionalities of the app is correct.

lsyurea committed:
Merge pull request #6 from lsyurea/branch-A-Assertions

Add assertions

lsyurea committed:
Fix checkstyle

Fixed the checkstyle of assertion statements.

lsyurea committed:
Update Parser.java

Reformatted if-else statement to remove else statement. This avoid deep nesting and make the happy path prominent.

lsyurea committed:
Merge pull request #7 from lsyurea/branch-A-CodeQuality

Branch a code quality

lsyurea committed:
Create RecurringTask.java

Created RecurringTask class to accept description and time.

lsyurea committed:
Add recurring feature

Updated parser to include recurring command. Updated RecurringTask to trim this.time in case it is stored wrongly.

lsyurea committed:
Update Main.java

Added Duke as the name of the application.

lsyurea committed:
Create Ui.png

Created the Ui photo.

lsyurea committed:
Update Ui.png

Changed to more realistic data.

lsyurea committed:
Update README.md

Added user guide using GFMD.

lsyurea committed:
Update README.md

Reformatted to make it more pleasing.

lsyurea committed:
Update README.md

lsyurea committed:
Update README.md

Added command summary.
